✨feat: 독서 모임 회원 관리 api 구현 완료

Yoon0221 opened the pull request:
## 🚀 변경사항
독서 모임 회원 관리 api 구현 완료

## 🔗 관련 이슈
- Closes #43 

## ✅ 체크리스트
- [x] 로컬에서 테스트 완료
- [x] 코드 리뷰 준비 완료

## 📝 특이사항
<!-- 리뷰어가 특별히 봐줬으면 하는 부분 (선택사항) -->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added API to allow users to apply for club membership.
  * Introduced API to retrieve club members with status filtering and cursor-based pagination.
  * Added API for administrators to update a club member's status or role.
  * Added API for members to leave a club.
  * Club member lists and responses now include staff role indicators and enhanced member details.

* **Improvements**
  * Enhanced error handling with new club membership-related error codes.
  * Improved data structures for club member responses, including updated fields for member and staff status.
  * Added support for managing club membership status updates and leaving clubs.
  * Improved category modification interfaces to use shared DTOs for consistency.
  * Enhanced pagination support for club member listings with standardized pageable parameters.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Yoon0221 committed:
✨feat: 독서 모임 가입 신청 api 구현 완료

Yoon0221 committed:
♻️refactor: clubMember - club 연관관계 설정

Yoon0221 committed:
✨feat: 독서 모임 회원 상태별 조회 api 구현 완료

Yoon0221 committed:
♻️refactor: 클럽 api - 해당 클럽의 운영진 여부 추가

Yoon0221 committed:
✨feat: 독서 모임 회원 상태 변경 api 구현 완료

Yoon0221 committed:
♻️refactor: 타겟 회원의 id 는 클럽 회원 id 로 받기

Yoon0221 committed:
♻️refactor: 주석 삭제

Yoon0221 committed:
✨feat: 독서 모임 회원 탈퇴 api 구현 완료

Yoon0221 committed:
♻️refactor: 로그인 기능 사용

Yoon0221 committed:
♻️refactor: coderabbitai 의견 반영

Yoon0221 committed:
♻️refactor: 잘못 추가함

MODUGGAGI commented on src/main/java/checkmo/domain/club/facade/ClubQueryFacadeImpl.java:
```diff
@@ -59,9 +60,44 @@ public ClubResponseDTO.ClubDetailDTO getClubInfo(Long clubId, String memberId) {
+        List<ClubResponseDTO.ClubMemberDTO> dtoList = members.stream()
+                .map(cm -> {
+                    MemberSharedDTO.BasicInfoDTO memberInfo = memberQueryFacade.getMemberBasicInfoForShare(cm.getMemberId());
+                    return new ClubResponseDTO.ClubMemberDTO(
```
ClubConverter 보니까 ClubResponseDTO.ClubMemberDTO로 반환하는 컨버터 있던데 혹시 이거 사용안하신 이유가 있을까요...??

Yoon0221 replied:
헉 몰랐습니다! 수정했어요~

MODUGGAGI commented on src/main/java/checkmo/domain/club/repository/ClubMemberRepository.java:
```diff
@@ -11,6 +11,14 @@
+
     @Query("SELECT c.id, c.name FROM ClubMember cm JOIN cm.club c WHERE cm.memberId = :memberId")
     List<Object[]> findClubIdAndNameByMemberId(@Param("memberId") String memberId);
+
```
이렇게 특정 개수 지정하는 것보다 파라미터로 Pageable pageable 받게 해놓고 개수는 PageRequest.of(0,size) 이런식으로 인자로 전달해서 조회하는 방식이 더 좋을 것 같아요

Yoon0221 replied:
넵 반영했습니다~ 감사합니다!

MODUGGAGI commented on src/main/java/checkmo/domain/club/service/command/ClubMembershipCommandServiceImpl.java:
```diff
@@ -0,0 +1,133 @@
+        ClubMember clubMember = ClubConverter.toClubMemberEntity(club, proxyMember, status, request.getJoinMessage());
+        club.addClubMember(clubMember);
+
+        return new ClubInfoDTO(clubId, null, club.isOpen());
```
얘도 그냥 저희 다른 사람들 코드 작성 형식이랑 맞추려면 builder 사용하는게 좋을것 같아요!!
그리구 그 저희가 기존에 주석보니까 가입 신청인 경우에는 클럽 이름에 null 넣자고 했는데, 그냥 이것도 일관성있게 클럽 이름 전달해주는게 좋을 것 같아요!

Yoon0221 replied:
넵 반영했습니다! 감사합니다~

zjhj0814 commented on src/main/java/checkmo/domain/club/converter/ClubConverter.java:
```diff
@@ -36,6 +36,49 @@ public class ClubConverter {
+     */
+    public static ClubResponseDTO.ClubMemberDTO toClubMemberDTO(ClubMember targetMember, MemberSharedDTO.BasicInfoDTO memberInfo) {
+        return ClubResponseDTO.ClubMemberDTO.builder()
+                .memberId(targetMember.getId())
```
여기 DTO 필드 이름을 ClubMemberId라고 하면 더 직관적일 것 같아요!

Yoon0221 replied:
넵 반영했습니다! 감사합니다~

zjhj0814 commented on src/main/java/checkmo/domain/club/converter/ClubConverter.java:
```diff
@@ -36,6 +36,49 @@ public class ClubConverter {
+        return ClubResponseDTO.ClubMemberDTO.builder()
+                .memberId(targetMember.getId())
+                .nickname(memberInfo.getNickname())
+                .profileImgUrl(memberInfo.getProfileImageUrl())
```
따로 MemberSharedDto에서 필드값을 분리하신 이유가 있을까요??

Yoon0221 replied:
오.. 질문이 잘 이해가 안갔어요! targetMember 랑 memberInfo 를 따로 쓰는 이유를 말씀하시는 건가요?
targetMember 는 클럽 도메인쪽 dto 라 클럽 멤버 관련 정보를 가져오고 memberInfo 는 멤버 쉐어 dto 라 멤버의 닉네임, 프로필 이미지를 따로 가져오고 있습니다! 
혹시 제가 잘못 이해했다면 말씀해주세요~!

MODUGGAGI replied:
파라미터에서 `MemberSharedDTO.BasicInfoDTO memberInfo`
얘는 내부 구성이
```
    @Getter
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class BasicInfoDTO {
        private String nickname;        // 회원 닉네임
        private String profileImageUrl; // 프로필 이미지 URL
    }
```
이렇게만 되어있는데,
```
                .nickname(memberInfo.getNickname())
                .profileImgUrl(memberInfo.getProfileImageUrl())
```
이렇게 따로 빼는거 말고 MemberSharedDTO.BasicInfoDTO로 묶는거 말씀하신 것 같아요!!

Yoon0221 replied:
아하 원래는 클럽 응답 DTO가 MemberSharedDTO에 직접 의존하지 않게 하려고 닉네임과 프로필 이미지를 각각 풀어서 넣었어요!
근데 말씀하신 대로 MemberSharedDTO.BasicInfoDTO 자체를 포함하면 더 간결해질 것 같아서 말씀하신대로 수정했습니다! 감사합니다~

zjhj0814 commented on src/main/java/checkmo/domain/club/facade/ClubQueryFacadeImpl.java:
```diff
@@ -59,9 +60,44 @@ public ClubResponseDTO.ClubDetailDTO getClubInfo(Long clubId, String memberId) {
+        // 3. DTO 변환
+        List<ClubResponseDTO.ClubMemberDTO> dtoList = members.stream()
+                .map(cm -> {
+                    MemberSharedDTO.BasicInfoDTO memberInfo = memberQueryFacade.getMemberBasicInfoForShare(cm.getMemberId());
```
각 사용자마다 기본 정보를 조회하는 로직이 반복되어서 MemberQueryFacade에 memberId 리스트로 배치 조회할 수 있는 메서드 만들어두었습니다!

지금까지 작성한 코드들은 리팩토링 단계에서 수정해도 상관없지만, 나중에 바꾸기 번거로우니 앞으로의 코드들은 배치 조회 사용해용!

Yoon0221 replied:
넵 감사합니다~ 반영했습니다!

zjhj0814 commented on src/main/java/checkmo/domain/club/repository/ClubMemberRepository.java:
```diff
@@ -11,6 +11,14 @@
+    List<ClubMember> findTop10ByClub_IdAndIdLessThanOrderByIdDesc(Long clubId, Long cursorId);
+
+    boolean existsByClub_IdAndClubMemberStatusAndIdLessThan(Long clubId, ClubMember.ClubMemberStatus status, Long lastId);
+    boolean existsByClub_IdAndIdLessThan(Long clubId, Long lastId);
```
다른 리포지토리 메서드와의 일관성을 위해샤 _ 지워야 할 것 가탕여!

Yoon0221 replied:
넵 반영했습니다~ 감사합니다!

MODUGGAGI replied:
이거 _ 지우는 거 반영이 안된 것 같아요..

Yoon0221 replied:
헉 정말이네요! 다시 수정했습니다! 죄송합니다아

Yoon0221 committed:
♻️refactor: 공유용 dto 가 아니라 내부 dto 를 공유함

Yoon0221 committed:
♻️refactor: 팀장님 피드백 반영 - Pageable pageable

Yoon0221 committed:
♻️refactor: 피드백 반영 - 코드 스타일 맞추기

Yoon0221 committed:
♻️refactor: 피드백 반영 - 코드 스타일 맞추기

Yoon0221 committed:
Merge branch 'develop' into feat/43/club-membership:wq!

Yoon0221 committed:
♻️refactor: 피드백 반영 - 배치 조회

MODUGGAGI commented on src/main/java/checkmo/domain/club/web/dto/club/ClubRequestDTO.java:
```diff
@@ -31,8 +31,8 @@ public static class ClubDetailDTO {
 
-        public CategoryRequestDTO.CategoryListRequestDTO toCategoryListRequestDTO() {
-            return CategoryRequestDTO.CategoryListRequestDTO.builder()
+        public CategorySharedDTO.CategoryIdListDTO toCategoryListRequestDTO() {
```
이거 일관성을 위해서 Converter로 옮기는 게 좋아보이는데... 혹시 여기에 두신 이유가 있을까요....?

Yoon0221 replied:
원래는 DTO 안에 변환 메서드를 두는 게 편리할 것 같아 넣었는데,
말씀하신 대로 Converter로 분리하는 게 훨씬 낫다고 생각해서 바로 수정했습니다!
좋은 지적 감사합니다~

MODUGGAGI commented on src/main/java/checkmo/domain/club/service/query/ClubQueryServiceImpl.java:
```diff
@@ -40,9 +44,65 @@ public ClubResponseDTO.MyClubListDTO getMyClubList(String memberId, int size) {
+        // 2. 조회 상태 변환
+        ClubMember.ClubMemberStatus clubMemberStatus = parseStatus(status);
+
+        if (clubMemberStatus == null) {
```
이 부분에 있는 if 절 내부의
- `clubMemberRepository.findByClubIdAndIdLessThanOrderByIdDesc(clubId, cursorId, pageable);`
- `clubMemberRepository.findByClubIdAndClubMemberStatusAndIdLessThanOrderByIdDesc(clubId, clubMemberStatus, cursorId, pageable);`

이 2개 부분에서 cursorId가 null인 첫페이지 조회에서 제대로 동작을 안할것 같아요...
아닌가요....? ㅠ

Yoon0221 replied:
헉 생각을 못했습니다! 
말씀해주신대로 cursorId가 null인 첫 페이지 조회 시 제대로 동작하지 않아서
현재는 cursorId null 여부에 따라 별도 쿼리를 호출하도록 분기 처리하여,
첫 페이지부터 정상적으로 조회되도록 수정해놓은 상태입니다!
좋은 지적 정말 감사합니다!

MODUGGAGI commented on src/main/java/checkmo/domain/club/web/controller/ClubController.java:
```diff
@@ -68,14 +71,108 @@ public ApiResponse<ClubResponseDTO.ClubDetailDTO> createClub(
+            @RequestParam(defaultValue = "ALL") String status, // 상태별 필터링 (MEMBER, STAFF, PENDING, BLOCKED, ALL 중 선택)
+            @RequestParam(required = false) Long cursorId // 페이징을 위한 커서 ID
+    ) {
+        Pageable pageable = PageRequest.of(0, 10);  // 페이지 0, 크기 10 고정
```
요기 페이징 처리할 때 사용할 변수인 Pageable 객체는 컨트롤러가 아니라 QueryFacade에서 처리하는 게 좋을 것 같아요!!
1. QueryFacade 내부에 상수를 선언해서 그 Pageable pageable = PageRequest.of(0, [상수]); 이런식으로 하는거랑, 
2. 아니면 컨트롤러에서 size도 쿼리 스트링으로 받아서 인자로 size 변수 넘긴 다음에 하는 방식

둘 중 한개로 해주시는게 좋을 것 같아요!!
저는 상수로 하는게 좀 더 깔끔할 것 같긴 해요

Yoon0221 replied:
넵 반영했습니다! 감사합니다~

MODUGGAGI commented on src/main/java/checkmo/domain/club/service/command/ClubMembershipCommandServiceImpl.java:
```diff
@@ -0,0 +1,137 @@
+        ClubMember clubMember = ClubConverter.toClubMemberEntity(club, proxyMember, status, request.getJoinMessage());
+        club.addClubMember(clubMember);
+
+        return ClubInfoDTO.builder()
```
요것도 간단한 거긴 한데 기존 코드와의 일관성을 위해서 컨버터에 넣어주시면 좋을 것 같아요!

Yoon0221 replied:
넵 반영했습니다~ 감사합니다!

MODUGGAGI commented on src/main/java/checkmo/domain/club/web/dto/club/ClubResponseDTO.java:
```diff
@@ -72,13 +72,24 @@ public static class ClubMemberListDTO {
     @AllArgsConstructor
     @Builder
     public static class ClubMemberDTO {
+        private Long clubMemberId; // 클럽 회원 ID
```
혹시 여기 DTO에 ClubMemberId가 추가되어야할 이유가 있을까요...?!

Yoon0221 replied:
지금 getClubMemberListByStatus에서 회원 목록을 조회해서 각 회원의 clubMemberId를 포함한 DTO를 반환하고,
이 clubMemberId를 사용해 updateClubMemberStatus API를 호출해서 특정 회원의 상태를 수정하고 있습니다!
getClubMemberListByStatus에서 해당 DTO 사용하고 있어서 추가했습니다!

Yoon0221 committed:
♻️refactor: 피드백 반영 - dto 묶기

Yoon0221 committed:
Merge branch 'develop' into feat/43/club-membership

Yoon0221 committed:
♻️refactor: 피드백 반영 - 리포지토리 name 형식 맞추기

Yoon0221 committed:
♻️refactor: 피드백 반영 - 컨버터 사용하기

Yoon0221 committed:
🔨fix: cursorId null 처리로 첫 페이지 조회 정상화

- cursorId가 null인 첫 페이지 조회 시 기존에 사용하던 IdLessThan 조건 쿼리가 정상 동작하지 않는 문제 수정
- cursorId null 여부에 따라 별도 쿼리 호출하도록 분기 처리하여 첫 페이지부터 정상 조회되도록 개선

Yoon0221 committed:
♻️refactor: 피드백 반영 - 페이징 크기 하드코딩 하지 말기

Yoon0221 committed:
♻️refactor: 피드백 반영 - 컨버터 사용하기
